src/daemon/distribution.c

```diff
@@ -0,0 +1,262 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Author:
+ *   Elene Margalitadze <elene.margalit at gmail.com>
+ **/
+
+#include "collectd.h"
+#include "distribution.h"
+#include <math.h>
+
+struct bucket_s {
+  uint64_t bucket_counter;
+  double min_boundary;
+  double max_boundary;
+};
+
+struct distribution_s {
+  bucket_t *buckets;
+  size_t num_buckets;
+  uint64_t total_scalar_count; // count of all registered scalar metrics
+  double raw_data_sum;         // sum of all registered raw scalar metrics
+};
+
+/*Private bucket constructor, min_boundary is inclusive, max_boundary is
+exclusive because the max_boundary Infinity is exclusive and we want the other
+max_boundaries to be consistent with that.*/
+static bucket_t initialize_bucket(double min_boundary, double max_boundary) {
+  bucket_t new_bucket = {
+      .bucket_counter = 0,
+      .min_boundary = min_boundary,
+      .max_boundary = max_boundary,
+  };
+  return new_bucket;
+}
+
+distribution_t *distribution_new_linear(size_t num_buckets, double size) {
+  if ((num_buckets == 0) || (size <= 0)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_buckets; i++) {
+    if (i < num_buckets - 1) {
+      new_distribution->buckets[i] =
+          initialize_bucket(i * size, i * size + size);
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(i * size, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets = num_buckets;
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+distribution_t *distribution_new_exponential(size_t num_buckets, double base,
+                                             double factor) {
+  if ((num_buckets == 0) || (base <= 0) || (factor <= 1)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  double previous_bucket_size;
+  double new_max;
+
+  for (size_t i = 0; i < num_buckets; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] = initialize_bucket(0, base);
+    } else if (i < num_buckets - 1) {
+      /*previous_bucket_size = new_distribution->buckets[i - 1].max_boundary -
```

Comment:
I'd recommend to remove code, rather than commenting it out. It feels scary, but you can get the old code back from the Git history if need be, so it's not lost.

Reply:
Okay.